crates/runtime/src/internal_table.rs

```diff
@@ -0,0 +1,140 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::sync::Arc;
+
+use arrow::record_batch::RecordBatch;
+use arrow::{array::ArrayRef, datatypes::Schema};
+use snafu::prelude::*;
+use tokio::sync::RwLock;
+
+use crate::accelerated_table::Retention;
+use crate::datafusion::DataFusion;
+use crate::datafusion::Error as DataFusionError;
+use crate::dataupdate::DataUpdate;
+use crate::{
+    accelerated_table::{refresh::Refresh, AcceleratedTable},
+    dataaccelerator::{self, create_accelerator_table},
+    dataconnector::{localhost::LocalhostConnector, DataConnector, DataConnectorError},
+};
+use spicepod::component::dataset::{acceleration::Acceleration, Dataset, Mode};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Unable to create data connector"))]
+    NoDataConnector {},
+
+    #[snafu(display("Unable to create data connector"))]
+    NoReadWriteProvider {},
+
+    #[snafu(display("Unable to create accelerated table provider"))]
+    UnableToCreateSourceTableProvider { source: DataConnectorError },
+
+    #[snafu(display("Unable to create accelerated table provider: {source}"))]
+    UnableToCreateAcceleratedTableProvider { source: dataaccelerator::Error },
+
+    #[snafu(display("Error creating record batch: {source}",))]
+    UnableToCreateRecordBatch { source: arrow::error::ArrowError },
+
+    #[snafu(display("Error writing to {name} table: {source}"))]
+    UnableToWriteToTable {
+        name: String,
+        source: DataFusionError,
+    },
+}
+
+pub struct InternalTable {
+    name: String,
+    schema: Arc<Schema>,
+    accelerated_table: Arc<AcceleratedTable>,
+}
+
+impl InternalTable {
+    #[must_use]
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    #[must_use]
+    pub fn accelerated_table(&self) -> Arc<AcceleratedTable> {
+        Arc::clone(&self.accelerated_table)
+    }
+
+    pub async fn insert(
+        &self,
+        data_fusion: &Arc<RwLock<DataFusion>>,
+        data: Vec<ArrayRef>,
+    ) -> Result<(), Error> {
+        let data_update = DataUpdate {
+            schema: Arc::clone(&self.schema),
+            data: vec![RecordBatch::try_new(Arc::clone(&self.schema), data)
+                .context(UnableToCreateRecordBatchSnafu)?],
+            update_type: crate::dataupdate::UpdateType::Append,
+        };
+
+        let df = data_fusion.write().await;
+        df.write_runtime_data(self.name(), data_update)
+            .await
+            .context(UnableToWriteToTableSnafu {
+                name: self.name().to_string(),
+            })?;
+        drop(df);
+
+        Ok(())
+    }
+
+    pub async fn new(
+        name: &str,
+        schema: Arc<Schema>,
+        acceleration: Acceleration,
+        refresh: Refresh,
+        retention: Option<Retention>,
+    ) -> Result<Self, Error> {
```

Comment:
Only localhost connector for now. As a next step I will add option to initialize internal table with spiceai connection